aten/src/ATen/native/cuda/Unique.cu

```diff
@@ -17,12 +22,100 @@ namespace native{
+  Tensor sorted;
+  const scalar_t* sorted_data = self_data;
+
+  Tensor inverse_indices;
+
+  if (return_inverse) {
+    inverse_indices = at::empty(self.sizes(), self.options().dtype(kLong));
+  }
+
+  if (!consecutive) {
+    sorted = at::empty({num_inp}, self.options());
+    scalar_t *sorted_data_ = sorted.data_ptr<scalar_t>();
+    sorted_data = sorted_data_;  // const pointer
+    if (return_inverse) {
+      auto arange = at::arange(0, num_inp, self.options().dtype(kLong));
+      int64_t *arange_data = arange.data_ptr<int64_t>();
+      sorted_indices_data = static_cast<int64_t *>(c10::cuda::CUDACachingAllocator::raw_alloc(num_inp * sizeof(int64_t)));
+      // get tmp_storage_size
+      cub::DeviceRadixSort::SortPairs(tmp_storage, tmp_storage_size, self_data, sorted_data_, arange_data, sorted_indices_data, num_inp, 0, sizeof(scalar_t) * 8, stream);
+      // allocate tmp_storage and run the computation
+      tmp_storage = c10::cuda::CUDACachingAllocator::raw_alloc(tmp_storage_size);
+      cub::DeviceRadixSort::SortPairs(tmp_storage, tmp_storage_size, self_data, sorted_data_, arange_data, sorted_indices_data, num_inp, 0, sizeof(scalar_t) * 8, stream);
+    } else {
+      // get tmp_storage_size
+      cub::DeviceRadixSort::SortKeys(tmp_storage, tmp_storage_size, self_data, sorted_data_, num_inp, 0, sizeof(scalar_t) * 8, stream);
+      // llocate tmp_storage and run the computation
+      tmp_storage = c10::cuda::CUDACachingAllocator::raw_alloc(tmp_storage_size);
+      cub::DeviceRadixSort::SortKeys(tmp_storage, tmp_storage_size, self_data, sorted_data_, num_inp, 0, sizeof(scalar_t) * 8, stream);
+    }
+  }
+
+  // inverse indices
+  if (return_inverse) {
+    auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+    auto policy = thrust::cuda::par(allocator).on(stream);
+    int64_t* inverse_indices_ptr = inverse_indices.data_ptr<int64_t>();
+    Tensor inv_loc = consecutive ? inverse_indices.view(-1) : at::empty({num_inp}, self.options().dtype(kLong));
+    int64_t* inv_loc_ptr = inv_loc.data_ptr<int64_t>();
+    thrust::adjacent_difference(policy, sorted_data, sorted_data + num_inp, inv_loc_ptr, thrust::not_equal_to<scalar_t>());
+    inv_loc[0] = 0;
+    thrust::inclusive_scan(policy, inv_loc_ptr, inv_loc_ptr + num_inp, inv_loc_ptr);
+    if (!consecutive) {
+      thrust::scatter(policy, inv_loc_ptr, inv_loc_ptr + num_inp, sorted_indices_data, inverse_indices_ptr);
+      c10::cuda::CUDACachingAllocator::raw_delete(sorted_indices_data);
+    }
+  }
+
+  // cub always returns counts, so just ignore the return_counts flag and return counts always
+  Tensor output = at::empty({num_inp}, self.options());
+  Tensor counts = at::empty({num_inp}, self.options().dtype(kLong));
+  scalar_t* output_data = output.data_ptr<scalar_t>();
+  int64_t* counts_data = counts.data_ptr<int64_t>();
+  auto num_out_tensor = at::empty({}, self.options().dtype(kInt));  // num_out used by cub is a device pointer
```

Comment:
it is possible to have more than 2**31 unique elements though? Especially for `consecutive`?

Reply:
It will fail I think. But FWIK `cub` always use `int` as the number of elements, this includes `cub::DeviceRunLengthEncode`  as well as  `cub::DeviceRadixSort`.
Should we:
1. Keep this PR as is (just assume the number of elements will be smaller than 2**31), this sounds bad, or
2. `TORCH_CHECK(num_inp < int_max, "unique does not support more than 2^32 elements");`, or
3. `if (num_inp > int_max) { thrust } else { cub }`